packages/dashboard/src/customization/widgets/lineSymbol/lineAnchor/component.tsx

```diff
@@ -0,0 +1,109 @@
+import React, { CSSProperties, useEffect, useRef } from 'react';
+import './component.css';
+import { LineWidget } from '~/customization/widgets/types';
+import { XYCoord, useDrag } from 'react-dnd';
+import { DashboardWidget } from '~/types';
+import { useGridSettings } from '~/components/actions/useGridSettings';
+import { useWidgetActions } from '~/customization/hooks/useWidgetActions';
+
+function computeNewPosition(
+  initialPosition: XYCoord,
+  offsetDifference: XYCoord,
+  height: number,
+  width: number,
+  cellSize: number
+) {
+  const { x: initialX, y: initialY } = initialPosition;
+  const { x: offsetX, y: offsetY } = offsetDifference;
+
+  const widthPixels = Math.round(width * cellSize);
+  const heightPixels = Math.round(height * cellSize);
+
+  const newX = Math.max(0, Math.min(initialX + offsetX, widthPixels));
+  const newY = Math.max(0, Math.min(initialY + offsetY, heightPixels));
+
+  return { x: newX, y: newY };
+}
+
+function getUpdatedWidget(
+  widget: LineWidget,
+  anchorType: 'start' | 'end',
+  initialPoint: XYCoord,
+  diff: XYCoord,
+  cellSize: number
+): DashboardWidget {
+  const { x, y } = computeNewPosition(initialPoint, diff, widget.height, widget.width, cellSize);
+  const updatedProperties = {
+    ...widget.properties,
+    [anchorType]: {
+      x,
+      y,
+    },
+  };
+
+  return {
+    ...widget,
+    properties: updatedProperties,
+  };
+}
+
+const useDragAndUpdate = (widget: LineWidget, anchorType: 'start' | 'end') => {
+  const { update: updateWidget } = useWidgetActions<DashboardWidget>();
+  const initialStartPointRef = useRef({
+    x: 0,
+    y: 0,
+  });
+  const initialEndPointRef = useRef({
+    x: 0,
+    y: 0,
+  });
+
+  const { cellSize } = useGridSettings();
+
+  const [{ diff, isDragging }, ref] = useDrag({
+    type: 'LineAnchor',
+    item: () => {
+      initialStartPointRef.current = {
+        x: widget.properties.start.x,
+        y: widget.properties.start.y,
+      };
+      initialEndPointRef.current = {
+        x: widget.properties.end.x,
+        y: widget.properties.end.y,
+      };
+      return {
+        anchorType,
+      };
+    },
+    collect: (monitor) => {
+      const offsetDifference = monitor.getDifferenceFromInitialOffset();
+      return { diff: offsetDifference, isDragging: monitor.isDragging() };
+    },
+  });
+
+  useEffect(() => {
+    if (isDragging && diff) {
+      const initialPoint = anchorType === 'start' ? initialStartPointRef.current : initialEndPointRef.current;
+      const updatedWidget = getUpdatedWidget(widget, anchorType, initialPoint, diff, cellSize);
+      updateWidget(updatedWidget);
+    }
+  }, [diff?.x, diff?.y]);
+  return ref;
+};
+
+export const LineAnchor: React.FC<{
+  style: CSSProperties;
+  anchorType: 'start' | 'end';
+  widget: LineWidget;
```

Comment:
does the anchor need the entire widget to function? I think it would be easier for it to just require a point

Reply:
the anchor uses `update` from `useWidgetActions`. we wouldn't be able to update the widget if we pass in only a point.

Reply:
Right, that is the correct way to update the widget, but it looks like the anchor is taking on too many responsibilities. For example, the hook initializes a ref for start and end point but only ever uses 1 of those. In memory this means we have extra refs that we don't need. It also handles updating the entire widget. I think it would be better if the line anchor only tracked 1 endpoint, and we can instead pass a function to update that endpoint on the widget. This would keep all the widget concerns in the line symbol component and the anchor concerns in the line anchor component.